metals/src/main/scala/scala/meta/internal/metals/codeactions/ExtractRenameMember.scala

```diff
@@ -0,0 +1,387 @@
+package scala.meta.internal.metals.codeactions
+
+import scala.collection.mutable.ListBuffer
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+
+import scala.meta.Defn
+import scala.meta.Import
+import scala.meta.Member
+import scala.meta.Mod
+import scala.meta.Pkg
+import scala.meta.Source
+import scala.meta.Template
+import scala.meta.Term
+import scala.meta.Tree
+import scala.meta.Type
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.metals._
+import scala.meta.internal.metals.codeactions.ExtractRenameMember.getMemberType
+import scala.meta.internal.parsing.Trees
+import scala.meta.io.AbsolutePath
+import scala.meta.pc.CancelToken
+import scala.meta.transversers.SimpleTraverser
+
+import org.eclipse.lsp4j.ApplyWorkspaceEditParams
+import org.eclipse.lsp4j.Location
+import org.eclipse.lsp4j.WorkspaceEdit
+import org.eclipse.{lsp4j => l}
+
+class ExtractRenameMember(
+    buffers: Buffers,
+    trees: Trees
+) extends CodeAction {
+
+  override def contribute(params: l.CodeActionParams, token: CancelToken)(
+      implicit ec: ExecutionContext
+  ): Future[Seq[l.CodeAction]] = Future {
+    val uri = params.getTextDocument.getUri
+    val path = uri.toAbsolutePath
+    val range = params.getRange
+
+    trees.get(path) match {
+      case Some(tree) =>
+        val fileName = uri.toAbsolutePath.filename.replaceAll("\\.scala$", "")
+
+        val definitions = membersDefinitions(tree)
+        val sldNames = sealedNames(tree)
+        val defnAtCursor =
+          definitions.find(_.name.pos.toLSP.overlapsWith(range))
+
+        def canRenameDefn(defn: Member): Boolean = {
+          val differentNames = defn.name.value != fileName
+          val notGetsExtendedByFileName =
```

Comment:
We should also probably check if the file already exists, even if it the rename would fail anyway I can see that it's better to not even give the code action option!
This check was introduced because when we were checking only classes, and not also traits/objects/enums, there could be a trait that had the same name as the file extended by a class. But now the `canRenameDefn` check is made only if there is exactly one member in the file, so all those checks are indeed useless. I will change them with the file exists! :)